File: app/src/main/java/com/ladsoft/bakingapp/mvp/presenter/StepPresenter.java
package com.ladsoft.bakingapp.mvp.presenter;

import com.ladsoft.bakingapp.entity.Step;
import com.ladsoft.bakingapp.mvp.StepsMvp;

import java.util.ArrayList;
import java.util.List;


public class StepPresenter extends Presenter<StepsMvp.View> {
    private List<Step> steps  = new ArrayList<>();
    private int stepIndex = 0;
    private boolean autoLoad = true;

    public void setData(int initialIndex, List<Step> steps, boolean autoLoad) {
        if (!isViewAttached()) { return; }

        if (steps != null) {
             stepIndex = initialIndex < steps.size()
                            ? initialIndex
                            : 0;

            this.steps.addAll(steps);

            if (autoLoad) {
                getView().onStepsLoaded(this.steps);
            } else {
                this.autoLoad = false;
            }
        }
    }

    public void showCurrentStep() {
        if (!isViewAttached()) { return; }

        if (steps.size() > 0) {
            if (!autoLoad) {
                getView().onStepsLoaded(steps);
                autoLoad = true;
            }

            getView().showStep(stepIndex);
        }
    }

    public void setCurrentStepIndex(int index) {
        stepIndex = index;
    }

    public void nextStep() {
        if (!isViewAttached()) { return; }

        if (stepIndex < this.steps.size()) {
            stepIndex = stepIndex + 1 == this.steps.size()
                            ? stepIndex
                            : stepIndex + 1;

            getView().showStep(stepIndex);
        }
    }

    public void previousStep() {
        if (!isViewAttached()) { return; }

        if (stepIndex > 0) {
            stepIndex = stepIndex == 0 ? stepIndex : stepIndex - 1;

            getView().showStep(stepIndex);
        }
    }

    public Step stepData() {
        return steps.get(stepIndex);
    }
}

File: app/src/main/java/com/ladsoft/bakingapp/mvp/presenter/StepPresenter.kt
package com.ladsoft.bakingapp.mvp.presenter

import com.ladsoft.bakingapp.entity.Step
import com.ladsoft.bakingapp.mvp.StepsMvp


open class StepPresenter : Presenter<StepsMvp.View>() {

    private val steps : MutableList<Step> = mutableListOf()
    private var stepIndex = 0
    private var autoLoad = true

    open fun setData(initialIndex: Int, steps: List<Step>?, autoLoad: Boolean = true) {
        if (steps != null) {
            when {
                initialIndex < steps.size -> stepIndex = initialIndex
                else -> stepIndex = 0
            }

            this.steps.addAll(steps)

            when {
                autoLoad -> view?.onStepsLoaded(this.steps)
                else -> this.autoLoad = false
            }
        }
    }

    open fun showCurrentStep() {
        steps.apply {
            if (size > 0) {
                if (!autoLoad) {
                    view?.onStepsLoaded(steps)
                    autoLoad = true
                }

                view?.showStep(stepIndex)
            }
        }
    }

    open fun setCurrentStepIndex(index: Int) {
        stepIndex = index
    }

    open fun nextStep() {
        if (stepIndex < this.steps.size) {
            stepIndex = if (stepIndex + 1 == this.steps.size) stepIndex else stepIndex + 1

            view?.showStep(stepIndex)
        }
    }

    open fun previousStep() {
        if (stepIndex > 0) {
            stepIndex = if (stepIndex == 0) stepIndex else stepIndex - 1

            view?.showStep(stepIndex)
        }
    }

    open fun stepData(): Step {
        return steps[stepIndex]
    }

}